lib/core/src/Cardano/Wallet/Network.hs

```diff
@@ -381,19 +490,182 @@ instance ToText FollowLog where
+flush tr t s = do
+    -- This is where we need to update the time
+    let s' = s { time = overCurrent (const t) (time s) }
+    traceWith tr s'
+    return $ hoistStats forgetPrev s'
+  where
+    forgetPrev (LogState _prev cur) = LogState cur cur
+
+emptyStats :: UTCTime -> SyncProgress -> FollowStats LogState
+emptyStats t p = FollowStats (f 0) (f 0) (f t) (f p)
+  where
+    f = initLogState
+
+-- | Update the stats based on a new log message.
+updateStats :: FollowLog tr -> FollowStats LogState -> FollowStats LogState
+updateStats msg s = case msg of
+    MsgApplyBlocks _tip blocks ->
+        s { blocksApplied = overCurrent (+ NE.length blocks) (blocksApplied s) }
+    MsgDidRollback _ _ ->
+        s { rollbacks = overCurrent (1 +) (rollbacks s) }
+    MsgSyncProgress p ->
+        s { prog = overCurrent (const p) (prog s) }
+    _ -> s
+
+instance ToText (FollowStats LogState) where
+    toText (FollowStats b r t sp) = syncStatus <> " " <> stats
+      where
+        -- NOTE: We are using NotResponding as initial value.
+        syncStatus = case sp of
+            LogState NotResponding Ready ->
+                "In sync."
+            LogState Ready Ready ->
+                "Still in sync."
+            LogState NotResponding NotResponding ->
+                "Still not syncing."
+            LogState (Syncing _p) Ready ->
+                "In sync!"
+            LogState Ready (Syncing p) ->
+                "Syncing again (" <> (pretty p) <> ")"
+            LogState _ (Syncing p) ->
+                "Syncing (" <> (pretty p) <> ")"
+            LogState prev NotResponding ->
+                "Not responding. Previously " <> (pretty prev) <> "."
+        stats = mconcat
+            [ "Applied " <> pretty (using (-) b) <> " blocks, "
+            , pretty (using (-) r) <> " rollbacks "
+            , "in the last " <> pretty (using diffUTCTime t) <> "."
+            ]
+          where
+            using f x = f (current x) (prev x)
+
+instance HasSeverityAnnotation (FollowStats LogState) where
+    getSeverityAnnotation s
+        | current (prog s) < prev (prog s) = Warning
+        | otherwise = Info
+        -- NOTE: Here we check if the sync progress is going backwards, which
+        -- would be a sign the wallet is overloaded (or rollbacks)
+        --
+        -- But this check might be in the wrong place. Might be better to
+        -- produce new logs from inside the updateStats function and immeditely
+        -- warn there.
+
+-- | Starts a new thread for monitoring health and statistics from
+-- the returned @FollowLog tr@.
+monitorFollowStats :: Tracer IO (FollowLog tr) -> IO (Tracer IO (FollowLog tr))
+monitorFollowStats tr = do
+    t0' <- getCurrentTime
+    let sp = NotResponding -- Hijacked as a default value. Should be fine.
+                           -- It appears to be otherwise unused in the codebase.
+    mvar <- newMVar $ emptyStats t0' sp
+    _ <- forkIO $ loop mvar (5 * second)
```

Comment:
`forkIO` — should presumable use something like `race` to get the monitoring thread to stop if `follow` stops.